src/rest/rest_utils.c

```diff
@@ -267,26 +724,88 @@ int database_prepare_array(json_t *j_array, linked_list_t *device_list)
 
     for (list_entry = device_list->head; list_entry != NULL; list_entry = list_entry->next)
     {
-        psk_string_len = sizeof(psk_string);
-        psk_id_string_len = sizeof(psk_id_string);
-
         device_entry = (database_entry_t *)list_entry->data;
 
-        base64_encode(device_entry->psk, device_entry->psk_len, psk_string, &psk_string_len);
-        base64_encode(device_entry->psk_id, device_entry->psk_id_len, psk_id_string, &psk_id_string_len);
+        memset(base64_secret_key, 0, sizeof(base64_secret_key));
+        memset(base64_public_key, 0, sizeof(base64_public_key));
+        memset(base64_serial, 0, sizeof(base64_serial));
+
+        base64_length = sizeof(base64_secret_key);
+        if (base64_encode(device_entry->secret_key, device_entry->secret_key_len, base64_secret_key, &base64_length))
+        {
+            return -1;
+        }
+
+        base64_length = sizeof(base64_public_key);
+        if (base64_encode(device_entry->public_key, device_entry->public_key_len, base64_public_key, &base64_length))
+        {
+            return -1;
+        }
+
+        base64_length = sizeof(base64_serial);
+        if (base64_encode(device_entry->serial, device_entry->serial_len, base64_serial, &base64_length))
+        {
+            return -1;
+        }
+
+        if (device_entry->mode == DEVICE_CREDENTIALS_PSK)
+        {
+            mode_string = "psk";
+        }
+        else if (device_entry->mode == DEVICE_CREDENTIALS_CERT)
+        {
+            mode_string = "cert";
+        }
+        else if (device_entry->mode == DEVICE_CREDENTIALS_NONE)
+        {
+            mode_string = "none";
+        }
+        else
+        {
+            return -1;
+        }
 
-        j_entry = json_pack("{s:s, s:s, s:s}", "uuid", device_entry->uuid, "psk", psk_string, "psk_id",
-                            psk_id_string);
+        j_entry = json_pack("{s:s, s:s, s:s, s:s, s:s, s:s}", "uuid", device_entry->uuid, "name", device_entry->name, "mode", mode_string, "secret_key", base64_secret_key, "public_key", base64_public_key, "serial", base64_serial);
         if (j_entry == NULL)
         {
             return -1;
         }
 
         if (json_array_append_new(j_array, j_entry))
         {
+            json_decref(j_entry);
             return -1;
         }
     }
 
     return 0;
 }
+
```

Comment:
as i see only `rest->settings->coap.certificate_file` is used, therefore scope could be narrowed.